Bluetooth: controller: Resolve AdvA in extended adv PDUs

andrzej-kaczmarek opened the pull request:
This allow to resolve AdvA in extended advertising PDUs, i.e. ADV_EXT_IND, AUX_ADV_IND, AUX_SCAN_RSP and AUX_CONNECT_RSP. Since offset of AdvA is different depending on PDU type (legacy vs.
extended) and PHY used (uncoded vs. coded), there's a bit more setup required.

When configuring AR, caller will provide a hint (channel and PHY) to radio driver on how to setup AR. The driver will assume extended PDU if non-1M or secondary channel is used and legacy PDU otherwise. Since it's possible that on 1M on primary we can still receive both legacy and extended PDUs, if the latter is received AR is triggered manually to resolve AdvA properly.

andrzej-kaczmarek committed:
Bluetooth: controller: Disable AAR PPI on AR reset

This was never disabled so would be triggered even if not explicitly
enabled.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Configure AR for resolving in ext adv

This patch allows to configure AR for resolving AdvA in extended adv
PDUs. The extra 'flags' field provides a hint from caller on how to
setup AR:
- if legacy PDU is expected, AdvA is set after PDU header,
- if extended PDU is expected, AdvA is set with +2B offset due to
  common extended advertising header,
- if Coded PHY is used, AdvA is set with extra +5b offset due to CI and
  TERM1 fields.

Note that AUX_SCAN_REQ and AUX_CONNECT_REQ are also considered legacy
PDUs here since they are exactly the same as SCAN_REQ and CONNECT_IND.

If neither legacy nor extended PDU is specified, it's derived from PHY,
i.e. 1M means legacy, non-1M means extended. Since it is still possible
to receive different type PDU in such case, it's up to LLL to detect
this and resolve AdvA when RX is done.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1104,24 +1143,47 @@ uint32_t radio_ccm_mic_is_valid(void)
-	hal_radio_nrf_ppi_channels_enable(BIT(HAL_TRIGGER_AAR_PPI));
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT)) {
+		nrf_radio_int_enable(NRF_RADIO, RADIO_INTENSET_BCMATCH_Msk);
```
Do not use BCMATCH ISR, for many reasons, more CPU use per PDU reception, dependency on CPU speed and OS ISR latencies to setup bit count compare in time, and higher peak current due to simultaneous radio-CPU use.

Thinking aloud, options, use the single NRF_AAR for either legacy PDU bit count compare or for common extended advertising payload format bit count compare. Have this as Kconfig. On Disabled ISR, use NRF_AAR again inline if the PPI based was incorrect. i.e. if PPI based was for legacy PDU and received PDU was extended, then inside the ISR resolve the address using NRF_AAR without IRQ (spin wait with low power, WFE).

andrzej-kaczmarek replied:
Expected PDU type can be implied from channel (pri/sec) and PHY. The only case where we can receive both types is 1M on primary so we need to use something by default. This could be configurable as Kconfig, although since I do not really see how one can decide which is better, for now I left this "hardcoded" to legacy and then resolve inline for extended. Adding a Kconfig should be pretty simple if needed.

andrzej-kaczmarek committed:
Bluetooth: controller: Add radio API to resolve addr via AAR

This allows to synchronously resolve any address using AAR. The AAR
has to be configured prior to calling this function.

The intended use for this function is to resolve AdvA in a PDU on
primary channel when AR was configured for different PDU type. That
means AR was configured to resolve in legacy PDU, but extended PDU
was received and we need to resolve AdvA manually. In such case, AAR
is already configured so just need to trigger it again with proper
ADDRPTR.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

andrzej-kaczmarek committed:
Bluetooth: controller: Resolve AdvA in ADV_EXT_IND properly

Since ADV_EXT_IND is received on primary channel, the AR is configured
to resolve AdvA in legacy PDU (assuming on 1M) and it will not resolve
AdvA properly. In such case, we should resolve AdvA manually in LLL.

Signed-off-by: Andrzej Kaczmarek <andrzej.kaczmarek@codecoup.pl>

cvinayak reviewed (approved):
QUICK-ACK.

cvinayak commented:
@anangl @rugeGerritsen bsim_hw_models is missing nrf_aar.h. Heard you guys are working on it?
@aescolar FYI

rugeGerritsen commented:
> @anangl @rugeGerritsen bsim_hw_models is missing nrf_aar.h. Heard you guys are working on it?
> @aescolar FYI

@cvinayak nrf_aar.h is already present. Maybe you need to rebase on master?
The HW models for the AAR were added long time back in https://github.com/BabbleSim/ext_NRF52_hw_models/commit/e9b643d6d5fa3be05b48b675d3856f9bdde327b4. The APIs were exposed not long ago in https://github.com/BabbleSim/ext_NRF52_hw_models/commit/d6ecc005a715119107a26f41ce899fb9892f8fc5.

cvinayak commented:
@andrzej-kaczmarek please rebase and force push

@rugeGerritsen after rebase 'nrf_aar_int_enable' is missing in nrf_hw_models' nrf_aar.h

rugeGerritsen commented:
@cvinayak, it seems like the simulator functionality is present but not exposed. To make it work, expose `nrf_aar_regw_sideeffects_INTENSET()` in NRF_AAR.h, add the following code to nrf_aar.c:
```
void nrf_aar_int_enable(NRF_AAR_Type * p_reg, uint32_t mask)
{
  p_reg->INTENSET = mask;
  nrf_aar_regw_sideeffects_INTENSET();
}
```
and expose `nrf_aar_int_enable()` in nrf_aar.h. The latter may be omitted if based upon https://github.com/BabbleSim/ext_NRF52_hw_models/pull/13

cvinayak commented:
PR for Babblesim: https://github.com/BabbleSim/ext_NRF52_hw_models/pull/14
Draft PR to check if it solves, yes. https://github.com/zephyrproject-rtos/zephyr/pull/27912
Draft PR failures are not related, commit message non-compliance and advx test failing (i suspect test timing sync between advx and scanx side)

andrzej-kaczmarek commented:
rebased & resolved conflicts

also handling of `flags` parameter in `radio_ar_configure` was changed since previously anything on non-1M was assumed as extended PDU. however, we can receive scan/connection request on non-1M which has the same payload as legacy PDU.

carlescufi commented:
@andrzej-kaczmarek there's a failure with the simulation that needs fixing: https://buildkite.com/zephyr/zephyr/builds/11160#3c8f7882-9736-4a27-96ee-a534b58eb542

andrzej-kaczmarek commented:
@carlescufi fixed
